Patch-mainline: v6.4-rc1
Git-commit: f8bb5104394560e29017c25bcade4c6b7aabd108
References: git-fixes
From: Wenliang Wang <wangwenliang.1995@bytedance.com>
Date: Thu, 4 May 2023 10:27:06 +0800
Subject: [PATCH] virtio_net: suppress cpu stall when free_unused_bufs

For multi-queue and large ring-size use case, the following error
occurred when free_unused_bufs:
rcu: INFO: rcu_sched self-detected stall on CPU.

Fixes: 986a4f4d452d ("virtio_net: multiqueue support")
Signed-off-by: Wenliang Wang <wangwenliang.1995@bytedance.com>
Acked-by: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Juergen Gross <jgross@suse.com>
---
 drivers/net/virtio_net.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 8d8038538fc4..a12ae26db0e2 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -2713,6 +2713,7 @@ static void free_unused_bufs(struct virtnet_info *vi)
 			else
 				xdp_return_frame(ptr_to_xdp(buf));
 		}
+		cond_resched();
 	}
 
 	for (i = 0; i < vi->max_queue_pairs; i++) {
@@ -2727,6 +2728,7 @@ static void free_unused_bufs(struct virtnet_info *vi)
 				put_page(virt_to_head_page(buf));
 			}
 		}
+		cond_resched();
 	}
 }
 
-- 
2.35.3

